File: lib/matestack/ui/vue_js/components/async.rb
module Matestack
  module Ui
    module VueJs
      module Components
        class Async < Matestack::Ui::VueJs::Vue
          vue_name 'matestack-ui-core-async'

          required :id
          optional :show_on, :hide_on, :rerender_on, :defer

          # register itself as an async component in the context
          def initialize(html_tag = nil, text = nil, options = {}, &block)
            super(html_tag, text, options, &block)
            Matestack::Ui::Core::Context.async_components = {} if Matestack::Ui::Core::Context.async_components.nil?
            Matestack::Ui::Core::Context.async_components[self.ctx.id] = self
          end

          def create_children(&block)
            self.response &block
          end

          def response
            if params[:component_key] && !is_not_requested?
              div async_attributes do
                yield
              end
            else
              vue_component do
                div class: 'matestack-async-component-container', 'v-bind:class': '{ "loading": loading === true }' do
                  div class: 'matestack-async-component-wrapper', 'v-if': 'asyncTemplate == null', 'v-bind:class': '{ "loading": loading === true }' do
                    div async_attributes do
                      if params[:component_key]
                        # we need to yield if a request is looking for a async component, indicated through present params[:component_key]
                        # the requested component could be hidden within this deferred async!
                        yield
                      else
                        yield unless is_deferred?
                      end
                    end
                  end
                  div class: 'matestack-async-component-wrapper', 'v-if': 'asyncTemplate != null', 'v-bind:class': '{ "loading": loading === true }' do
                    Matestack::Ui::Core::Base.new('v-runtime-template', ':template': 'asyncTemplate')
                  end
                end
              end
            end
          end

          def async_attributes
            options.merge({
              id: ctx.id, 
              class: 'matestack-async-component-root', 
              'v-if': 'showing'
            })
          end

          def config
            {
              component_key: ctx.id,
              show_on: ctx.show_on,
              hide_on: ctx.hide_on,
              rerender_on: ctx.rerender_on,
              defer: ctx.defer,
              parent_class: isolated_parent ? isolated_parent.class.to_s : nil,
            }
          end

          def is_deferred?
            ctx.defer
          end

          def is_not_requested?
            params[:component_key].present? && params[:component_key] != ctx.id
          end

          def isolated_parent
            Matestack::Ui::Core::Context.isolated_parent
          end

        end
      end
    end
  end
end

File: lib/matestack/ui/vue_js/components/async.js
import Vue from 'vue/dist/vue.esm'
import axios from 'axios'
import VRuntimeTemplate from "v-runtime-template"
import matestackEventHub from '../event_hub'
import componentMixin from './mixin'

const componentDef = {
 mixins: [componentMixin],
 data: function(){
   return {
     asyncTemplate: null,
     showing: true,
     loading: false,
     hideAfterTimeout: null,
     event: {
       data: {}
     }
   }
 },
 methods: {
   show: function(event_data){
     const self = this
     if (this.showing === true){
       return
     }
     this.showing = true
     this.event.data = event_data
     if(this.props["defer"] != undefined){
       if(!isNaN(this.props["defer"])){
         this.startDefer()
       }
     }
     if(this.props["hide_after"] != undefined){
       self.hideAfterTimeout = setTimeout(function () {
         self.hide()
       }, parseInt(this.props["hide_after"]));
     }
   },
   hide: function(){
     this.showing = false
     this.event.data = {}
   },
   startDefer: function(){
     const self = this
     self.loading = true;
     setTimeout(function () {
       self.rerender()
     }, parseInt(this.props["defer"]));
   },
   rerender: function(){
     var self = this;
     self.loading = true;
     axios({
       method: "get",
       url: location.pathname + location.search,
       headers: {
         'X-CSRF-Token': document.getElementsByName("csrf-token")[0].getAttribute('content')
       },
       params: {
         "component_key": self.props["component_key"],
         "component_class": self.props["parent_class"]
       }
     })
         .then(function(response){
           var tmp_dom_element = document.createElement('div');
           tmp_dom_element.innerHTML = response['data'];
           var template = tmp_dom_element.querySelector('#' + self.props["component_key"]).outerHTML;
           self.loading = false;
           self.asyncTemplate = template;
         })
         .catch(function(error){
           console.log(error)
           matestackEventHub.$emit('async_rerender_error', { id: self.props["component_key"] })
         })
   }
 },
 created: function () {
   const self = this
   self.registerEvents(this.props['show_on'], self.show)
   self.registerEvents(this.props['hide_on'], self.hide)
   self.registerEvents(this.props['rerender_on'], self.rerender)
   if(this.props["show_on"] != undefined){
     this.showing = false
   }
   if(this.props["defer"] != undefined){
     if(!isNaN(this.props["defer"])){
       if (this.props["show_on"] == undefined){
         this.startDefer()
       }
     }
   }
   if(this.props["init_show"] == true){
     console.log("(inside the if init show = true - it's showing !!!!!")
     this.showing = true
   }
 },
 beforeDestroy: function() {
   const self = this
   clearTimeout(self.hideAfterTimeout)
   puts("%%%%% hi ")
   self.removeEvents(this.props["show_on"], self.show)
   self.removeEvents(this.props["hide_on"], self.hide)
   self.removeEvents(this.props["rerender_on"], self.rerender)
 },
 components: {
   VRuntimeTemplate: VRuntimeTemplate
 }
}

let component = Vue.component('matestack-ui-core-async', componentDef)

export default componentDef

